src/Compilers/CSharp/Portable/Symbols/Synthesized/InjectedNonNullTypesAttributeSymbol.cs

```diff
@@ -0,0 +1,172 @@
+                cancellationToken.ThrowIfCancellationRequested();
+                var incompletePart = _state.NextIncompletePart;
+                switch (incompletePart)
+                {
+                    case CompletionPart.Attributes:
+                        GetAttributes();
+                        break;
+
+                    case CompletionPart.None:
+                        return;
+
+                    default:
+                        // any other values are completion parts intended for other kinds of symbols
+                        _state.NotePartComplete(CompletionPart.All & ~CompletionPart.InjectedSymbolAll);
+                        break;
+                }
+
+                _state.SpinWaitComplete(incompletePart, cancellationToken);
+            }
+
+            throw ExceptionUtilities.Unreachable;
+        }
+
+        public override ImmutableArray<CSharpAttributeData> GetAttributes()
+        {
+            var attributes = _lazyCustomAttributes;
+            if (!attributes.IsDefault)
+            {
+                return attributes;
+            }
+
+            if (ImmutableInterlocked.InterlockedInitialize(ref _lazyCustomAttributes, MakeAttributes()))
+            {
+                var completed = _state.NotePartComplete(CompletionPart.Attributes);
+                Debug.Assert(completed);
+            }
+
+            return _lazyCustomAttributes;
+        }
+
+        /// <summary>
+        /// Adds an `[AttributeUsage(AttributeTargets.Class | ...)]` (if possible) and captures any diagnostics in the process.
+        /// </summary>
+        private ImmutableArray<CSharpAttributeData> MakeAttributes()
+        {
+            var ctor = (MethodSymbol)DeclaringCompilation.GetWellKnownTypeMember(WellKnownMember.System_AttributeUsageAttribute__ctor);
+            if (ctor is null)
+            {
+                // member is missing
+                var memberDescriptor = WellKnownMembers.GetDescriptor(WellKnownMember.System_AttributeUsageAttribute__ctor);
+                var diagnosticInfo = new CSDiagnosticInfo(ErrorCode.ERR_MissingPredefinedMember, memberDescriptor.DeclaringTypeMetadataName, memberDescriptor.Name);
+                Diagnostics.Add(diagnosticInfo, Location.None);
+
+                return ImmutableArray<CSharpAttributeData>.Empty;
+            }
+
+            Binder.ReportUseSiteDiagnostics(ctor, Diagnostics, Location.None);
+
+            NamedTypeSymbol attributeTargets = DeclaringCompilation.GetWellKnownType(WellKnownType.System_AttributeTargets);
+            Binder.ReportUseSiteDiagnostics(attributeTargets, Diagnostics, Location.None);
+
+            var usage = new SynthesizedAttributeData(ctor,
+                arguments: ImmutableArray.Create(new TypedConstant(attributeTargets, TypedConstantKind.Enum, attributeUsage)),
+                namedArguments: ImmutableArray<KeyValuePair<string, TypedConstant>>.Empty);
+
+            return ImmutableArray.Create<CSharpAttributeData>(usage);
```

Comment:
>ImmutableArray.Create<CSharpAttributeData>(usage) [](start = 19, length = 49)

Should we do the same for an Embedded attribute?

Reply:
The Embedded attribute cannot be referenced in source, so any usage gets an error. I didn't see a benefit to documenting an allowed targets list (the default seems fine).

Reply:
> The Embedded attribute cannot be referenced in source, so any usage gets an error. I didn't see a benefit to documenting an allowed targets list (the default seems fine).

What was the behavior prior to this change?

---
In reply to: [215651712](https://github.com/dotnet/roslyn/pull/29180#discussion_r215651712) [](ancestors = 215651712)

Reply:
Prior to this PR:
- If the attribute is synthesized, then you cannot use it in source and it is generated without AttributeUsage. ([example](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAGAAhwEYBuAWACgcBmIgJgIGEAeAFQD4CB3AC0TgE2IAgFcAdpHEQA5nAAmBAN5UCagqo2UAvlSpA===))
- If the attribute exists in source, then the user can put whatever AttributeUsage they want.



---
In reply to: [215706141](https://github.com/dotnet/roslyn/pull/29180#discussion_r215706141) [](ancestors = 215706141,215651712)